test/suite/wthooks.py

```diff
@@ -0,0 +1,216 @@
+
+            # If we're using the wiredtiger module and not a class, we need a slightly different style of hooked_function.
+            if clazz == wiredtiger:
+                f = lambda *args: hooked_function(wiredtiger, orig_func, hooks_name, *args)
+            else:
+                f = lambda self, *args: hooked_function(self, orig_func, hooks_name, *args)
+            setattr(clazz, method_name, f)
+            setattr(clazz, orig_name, orig_func)
+
+        # Now add to the list of hook functions
+        # If it's a replace hook, we only allow one of them for a given method name
+        hooks_list = getattr(clazz, hooks_name)
+        if hook_type == HOOK_REPLACE:
+            for ht, hf in hooks_list:
+                if ht == HOOK_REPLACE:
+                    raise Exception('method ' + method_name + ' hook setup: trying to replace the same method with two hooks')
+        elif hook_type != HOOK_NOTIFY and hook_type != HOOK_ARGS:
+            raise Exception('method ' + method_name + ' hook setup: unknown hook_type: ' + str(hook_type))
+        hooks_list.append([hook_type, hook_func])
+        setattr(clazz, hooks_name, hooks_list)
+        tty('Setting up hooks list in ' + str(clazz) + '.' + hooks_name)
+
+    def get_function(self, clazz, method_name):
+        orig_name = '_' + method_name + '_orig'
+        if hasattr(clazz, orig_name):
+            orig_func = getattr(clazz, orig_name)
+        else:
+            orig_func = getattr(clazz, method_name)
+        return orig_func
+
+    def filter_tests(self, tests):
+        for hook in self.hooks:
+            tests = hook.filter_tests(tests)
+        return tests
+
+# Hooks must derive from this class
+class WiredTigerHook(object):
+    def __init__(self):
+        pass
+
+    # This is called right after creation and should not be overridden.
+    def _initialize(self, name, hookmgr):
+        self.name = name
+        self.hookmgr = hookmgr
+
+    # default version of filter_tests, can be overridden
+    def filter_tests(self, tests):
+        return tests
+
+    def setup_hooks(self):
+        raise Exception('setup_hooks must be overridden by a hook class')
+
+    # Call these to do override of "global" WiredTiger functions, like wiredtiger_open.
+    def add_wiredtiger_hook(self, method_name, hook_type, hook_func):
+        self.hookmgr.add_hook(wiredtiger, method_name, hook_type, hook_func)
+
+    def add_connection_hook(self, method_name, hook_type, hook_func):
+        self.hookmgr.add_hook(wiredtiger.Connection, method_name, hook_type, hook_func)
+
+    def add_session_hook(self, method_name, hook_type, hook_func):
+        self.hookmgr.add_hook(wiredtiger.Session, method_name, hook_type, hook_func)
+
+    def add_cursor_hook(self, method_name, hook_type, hook_func):
+        self.hookmgr.add_hook(wiredtiger.Cursor, method_name, hook_type, hook_func)
+
+    # Allows lookup of the original version of functions
+    def get_wiredtiger_function(self, method_name):
```

Comment:
Would merging all of these into one getter and setter function be more easier to understand? Applying something like adding **object_name** into the method, and grabbing the object via wiredtiger[object_name].

Reply:
I took a stab at this - not quite sure your vision of this was.  Here's what I came up with:
```
            orig_session_create = self.Session['create']
            self.Session['open_cursor'] = (wthooks.HOOK_NOTIFY, session_open_cursor_notify)
```
I'm using `self.Session` rather than `Session`, because I was a little uncomfortable overriding the setitem/getitem on the `Connection/Session/Cursor` class or `wiredtiger` module.  `self.Session` is a proxy object that gets the same thing done.

The assignment needs two arguments, hence a tuple.

It's _slightly_ unintuitive in that an "assignment" adds a hook, but doesn't override any existing hook, unlike regular assignment.  We could also do something like:
```
       self.Session['open_cursor'].append(wthooks.HOOK_NOTIFY, session_open_cursor_notify)
```

Reply:
Sorry for the misunderstanding, my initial vision was to combine the multiple functions through something like:
**
def get_connection_function(self, method_name, object_name=None):
    if object_name != None and hasattr(wiredtiger, object_name):
        return self.hookmgr.get_function(wiredtiger[object_name], method_name)
    else:
        return self.hookmgr.get_function(wiredtiger, method_name)
** 
Although this version works well too, only issue might be the confusion when assigning more functions to a particular module e.g. wiredtiger, as it actually represents an **additem**, and adding the complexity of the proxy class.